src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,474 @@
+            "PaymentCoordinator._payForRange: startTimestamp too far in the future"
+        );
+
+        // Require rangePayment is for whitelisted strategy or beaconChainETHStrategy
+        address currAddress = address(0);
+        for (uint256 i = 0; i < rangePayment.strategiesAndMultipliers.length; ++i) {
+            IStrategy strategy = rangePayment.strategiesAndMultipliers[i].strategy;
+            require(
+                strategyManager.strategyIsWhitelistedForDeposit(strategy) ||
+                    strategy == beaconChainETHStrategy,
+                "PaymentCoordinator._payForRange: invalid strategy considered"
+            );
+            require(
+                currAddress < address(strategy),
+                "PaymentCoordinator._payForRange: strategies must be in ascending order to handle duplicates"
+            );
+            currAddress = address(strategy);
+        }
+
+        rangePayment.token.safeTransferFrom(msg.sender, address(this), rangePayment.amount);
+    }
+
+    function _checkClaim(PaymentMerkleClaim calldata claim, DistributionRoot memory root) internal view returns (bool) {
+        require(block.timestamp >= root.activatedAt, "PaymentCoordinator._checkClaim: root not activated yet");
+        require(
+            claim.tokenIndices.length == claim.tokenTreeProofs.length,
+            "PaymentCoordinator._checkClaim: tokenIndices and tokenProofs length mismatch"
+        );
+        require(
+            claim.tokenTreeProofs.length == claim.tokenLeaves.length,
+            "PaymentCoordinator._checkClaim: tokenTreeProofs and leaves length mismatch"
+        );
+
+        // Verify inclusion of earners leaf (earner, earnerTokenRoot) in the distribution root
+        _verifyEarnerClaimProof({
+            root: root.root,
+            earnerLeafIndex: claim.earnerIndex,
+            earnerProof: claim.earnerTreeProof,
+            earnerLeaf: claim.earnerLeaf
+        });
+        // For each of the tokenLeaf proofs, verify inclusion of token tree leaf again the earnerTokenRoot
+        for (uint256 i = 0; i < claim.tokenIndices.length; ++i) {
+            _verifyTokenClaimProof({
+                earnerTokenRoot: claim.earnerLeaf.earnerTokenRoot,
+                tokenLeafIndex: claim.tokenIndices[i],
+                tokenProof: claim.tokenTreeProofs[i],
+                tokenLeaf: claim.tokenLeaves[i]
+            });
+        }
+
+        return true;
+    }
+
+    /**
+     * @notice verify inclusion of the token claim proof in the earner token root hash (earnerTokenRoot).
+     * The token leaf comprises of the IERC20 token and cumulativeAmount of earnings.
+     * @param earnerTokenRoot root hash of the earner token subtree
+     * @param tokenLeafIndex index of the token leaf
+     * @param tokenProof proof of the token leaf in the earner token subtree
+     * @param tokenLeaf token leaf to be verified
+     */
+    function _verifyTokenClaimProof(
```

Comment:
Because we don't validate the size of `tokenLeafIndex` relative to `tokenProof`, there are multiple indices that are valid for any single given proof.

We're not using the indices to check whether a leaf has been claimed or not, but we should probably adhere to best practices anyway!

IIRC, indices should meet this criterion: `index < 2**(proof.length/32)`

(this applies to the other merkle proof method below)

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/514/commits/24fb70726a39a125e61cc19a361d7cc20aa7b930
Added checks and tests for both verifyToken and verifyEarner

Reply:
Used bit shifting for cheaper gas